### PR TITLE
Example minimize evaluation utility.

### PR DESCRIPTION
**Pull request by ajreynol**
This implements a simple algorithm for minimizing the number of calls to the evaluator.

It is work towards optimizing our performance on "examples" conjectures (from this branch: https://github.com/ajreynol/CVC4/tree/optExamplesSygus).

### CODE REVIEW

File: src/theory/quantifiers/sygus/example_min_eval.h [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+   */
+  Node evaluate(const std::vector<Node>& subs);
+
+ private:
+  /** The node to evaluate */
+  Node d_evalNode;
+  /** The domain of substitutions */
+  std::vector<Node> d_vars;
+  /** The indices in d_vars that occur free in n */
+  std::vector<unsigned> d_indices;
```
Comment (4tXJ7f) [inline src/theory/quantifiers/sygus/example_min_eval.h]:
```suggestion
  std::vector<size_t> d_indices;
```

File: src/theory/quantifiers/sygus/example_min_eval.cpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                                EmeEval* ece)
+{
+  AlwaysAssert(d_evalNode.isNull());
+  d_evalNode = n;
+  d_vars.insert(d_vars.end(), vars.begin(), vars.end());
+
+  // compute its free variables
+  std::unordered_set<Node, NodeHashFunction> fvs;
+  expr::getFreeVariables(n, fvs);
+  for (unsigned i = 0, vsize = vars.size(); i < vsize; i++)
```
Comment (4tXJ7f) [inline src/theory/quantifiers/sygus/example_min_eval.cpp]:
```suggestion
  for (size_t i = 0, vsize = vars.size(); i < vsize; i++)
```

File: src/theory/quantifiers/sygus/example_min_eval.cpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+                         << d_vars.size() << " variables are relevant"
+                         << std::endl;
+  d_ece = ece;
+}
+
+Node ExampleMinEval::evaluate(const std::vector<Node>& subs)
+{
+  Assert(d_vars.size() == subs.size());
+  
+  if (d_indices.size()==d_vars.size())
```
Comment (4tXJ7f) [inline src/theory/quantifiers/sygus/example_min_eval.cpp]: Apply Clang-Format.

File: src/theory/quantifiers/sygus/example_min_eval.h [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+class ExampleMinEval
+{
+ public:
+  ExampleMinEval() {}
+  ~ExampleMinEval() {}
+  /**
+   * Initialize this cache to evaluate n on substitutions with domain vars.
+   * Argument ece is the evaluator object.
+   */
+  void initialize(Node n, const std::vector<Node>& vars, EmeEval* ece);
```
Comment (4tXJ7f) [inline src/theory/quantifiers/sygus/example_min_eval.h]: Not a huge fan of this empty constructor & initialize pattern as you know :) Is this really needed?

Comment (ajreynol) [reply src/theory/quantifiers/sygus/example_min_eval.h]: This one isn't, I've updated the code.

**Pull request merged**